### PR TITLE
fix: make nulls-order consistent with postgres

### LINKED ISSUE #1343: NULLS ORDER is inconsistent with postgres
**Describe the bug**
https://github.com/apache/arrow-datafusion/pull/1339#discussion_r753655676

**To Reproduce**

**Expected behavior**

**Additional context**
https://www.postgresql.org/docs/current/queries-order.html

### PR DESCRIPTION
**Pull request by xudong963**
# Which issue does this PR close?

Closes #1343 

 # Rationale for this change

# What changes are included in this PR?

# Are there any user-facing changes?

No

### CODE REVIEW

File: datafusion/tests/sql.rs [hunk 1]
```diff
@@ -5999,3 +5999,75 @@ async fn test_expect_distinct() -> Result<()> {
     assert_batches_eq!(expected, &actual);
     Ok(())
 }
+
+#[tokio::test]
+async fn test_nulls_first_asc() -> Result<()> {
+    let mut ctx = ExecutionContext::new();
+    let sql = "SELECT * FROM (VALUES (1, 'one'), (2, 'two'), (null, 'three')) AS t (num,letter) ORDER BY num";
```
Comment (xudong963) [inline datafusion/tests/sql.rs:6006]: Every time I construct a test case with a **value list**, I think of @Jimexist lol 😄

**Review by Dandandan — APPROVED** [review]

**Comment by Dandandan** [comment]
Might be good to add some tests as well to compare it with PostgreSQL?

**Comment by xudong963** [comment]
> Might be good to add some tests as well to compare it with PostgreSQL (integration-tests folder)?

OK

**Review by alamb — APPROVED** [review]
I tried it out with postgres and indeed this appears consistent

```
alamb=# select * from A order by a;
 a 
---
 1
 2
  
(3 rows)
```

I agree with @Dandandan  that some integration tests would be nice, though probably not necessary in this case

**Comment by xudong963** [comment]
> I tried it out with postgres and indeed this appears consistent
> 
> ```
> alamb=# select * from A order by a;
>  a 
> ---
>  1
>  2
>   
> (3 rows)
> ```
> 
> I agree with @Dandandan that some integration tests would be nice, though probably not necessary in this case

Done, PTAL! cc @Dandandan

File: datafusion/tests/sql.rs (hunk 1 above)
Comment (alamb) [reply datafusion/tests/sql.rs:6006]: yeah it is so handy

File: datafusion/tests/sql.rs [hunk 2]
```diff
@@ -5999,3 +5999,75 @@ async fn test_expect_distinct() -> Result<()> {
...
+    let expected = vec![
+        "+-----+--------+",
+        "| num | letter |",
+        "+-----+--------+",
+        "|     | three  |",
+        "| 2   | two    |",
+        "| 1   | one    |",
+        "+-----+--------+",
+    ];
+    assert_batches_eq!(expected, &actual);
```
Comment (alamb) [inline datafusion/tests/sql.rs:6035]: 👍

**Review by alamb — APPROVED** [review]
Looks great to me. Thanks @xudong963

**Review by houqp — APPROVED** [review]

**Comment by Dandandan** [comment]
Thanks @xudong963 !

**Comment by alamb** [comment]
🎉

**Pull request merged**